hedera-node/hedera-networkadmin-service-impl/src/main/java/com/hedera/node/app/service/networkadmin/impl/handlers/UnzipUtility.java

```diff
@@ -0,0 +1,137 @@
+
+    private static final int BUFFER_SIZE = 4096;
+
+    // these constants are used to prevent Zip Bomb Attacks
+    private static final int THRESHOLD_ENTRIES = 10000; // max # of entries to allow in zip files
+    private static final int THRESHOLD_ZIP_SIZE = 1000000000; // 1 GB - max allowed total size of all uncompressed files
+    private static final int THRESHOLD_ENTRY_SIZE = 100000000; // 100 MB - max allowed size of one uncompressed file
+    // max allowed ratio between uncompressed and compressed file size
+    private static final double THRESHOLD_RATIO = 10;
+
+    private UnzipUtility() {
+        throw new UnsupportedOperationException("Utility Class");
+    }
+
+    /**
+     * Extracts (unzips) a zipped file from a byte array
+     * @param bytes the byte array containing the zipped file
+     * @param dstDir the destination directory to extract the unzipped file to
+     * @throws IOException if the destination does not exist and can't be created, or if the file can't be written
+     */
+    public static void unzip(@NonNull final byte[] bytes, @NonNull final Path dstDir) throws IOException {
+        requireNonNull(bytes);
+        requireNonNull(dstDir);
+
+        final var zipIn = new ZipInputStream(new ByteArrayInputStream(bytes));
+        int totalSizeArchive = 0; // total size of the zip archive once uncompressed
+        int totalEntryArchive = 0; // total number of entries in the zip archive
+        ZipEntry entry = zipIn.getNextEntry();
+
+        if (entry == null) {
+            throw new IOException("No zip entry found in bytes");
+        }
+        while (entry != null) {
+            totalEntryArchive++;
+            if (totalEntryArchive > THRESHOLD_ENTRIES) {
+                throw new IOException("Zip file entry count exceeds threshold: " + THRESHOLD_ENTRIES);
+            }
+            Path filePath = dstDir.resolve(entry.getName());
+            final File fileOrDir = filePath.toFile();
+            final File directory = fileOrDir.getParentFile();
+            if (!directory.exists() && !directory.mkdirs()) {
+                throw new IOException("Unable to create the parent directories for the file: " + fileOrDir);
+            }
+
+            if (!entry.isDirectory()) {
+                totalSizeArchive += extractSingleFile(zipIn, filePath, entry.getCompressedSize());
+                if (totalSizeArchive > THRESHOLD_ZIP_SIZE) {
+                    throw new IOException("Zip file size exceeds threshold: " + THRESHOLD_ZIP_SIZE);
+                }
+                log.info(" - Extracted update file {}", filePath);
+            } else {
+                if (!fileOrDir.mkdirs()) {
+                    throw new IOException("Unable to create assets sub-directory: " + fileOrDir);
+                }
+                log.info(" - Created assets sub-directory {}", fileOrDir);
+            }
+            zipIn.closeEntry();
+            entry = zipIn.getNextEntry();
+        }
+        zipIn.close();
```

Comment:
Does the ZIP API supports try-with-ressource? In that case it would make sense to use to call the close methods in case of exceptions.

Reply:
Nice catch. Fixed.